File: src/main/java/com/kotoriSystems/controller/are_doko/ItemController.java
package com.kotoriSystems.controller.are_doko;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

import com.kotoriSystems.entity.are_doko.Item;
import com.kotoriSystems.model.SearchForm;
import com.kotoriSystems.repository.are_doko.ItemRepository;

@Controller
public class ItemController {

  @Autowired
  private ItemRepository itemRepository;

  // @Autowired
  // private LocationRepository locationRepository;

  // @Autowired
  // private GenreRepository genreRepository;

  @GetMapping("/items/new")
  public String addItem(@ModelAttribute Item item) {
    return "item_form";
  }

  @PostMapping("/items/save")
  public String saveItem(@ModelAttribute Item item, RedirectAttributes redirectAttributes) {
    try {
      System.out.println("testtest saveItem:" + item.getId() + "," + item.getName() + "," + item.getLocation().getName() + ","
          + item.getLocationDetail() + "," + item.getGenre().getName());
      itemRepository.save(item);

      redirectAttributes.addFlashAttribute("message", "The Item has been saved successfully!");
    } catch (Exception e) {
      redirectAttributes.addAttribute("message", e.getMessage());
    }

    return "redirect:items";
  }

  @GetMapping("/items")
  public String listItem(@ModelAttribute SearchForm form) {
    List<Item> items = itemRepository.findAll();

    for (Item item : items) {
      System.out.println("testtest item:" + item.getId() + "," + item.getName() + "," + item.getLocation().getName() + ","
          + item.getLocationDetail() + "," + item.getGenre().getName());
    }

    form.setItems(items);
    return "are_doko/home";
  }

  @GetMapping("/items/{id}")
  public String updateItem(@PathVariable("id") long id, Model model,
      RedirectAttributes redirectAttributes) {
    try {
      // idに対応するItemを取得
      Item existingItem = itemRepository.findById(id)
          .orElseThrow(() -> new IllegalArgumentException("Invalid item Id:" + id));
      System.out.println("testtest existingItem:" + existingItem.getId() + "," + existingItem.getName() + "," + existingItem.getLocation().getName() + ","
      + existingItem.getLocationDetail() + "," + existingItem.getGenre().getName());

      model.addAttribute("item", existingItem);
      model.addAttribute("pageTitle", "Edit Item (ID: " + id + ")");

      return "are_doko/home_form";
    } catch (Exception e) {
      redirectAttributes.addFlashAttribute("message", e.getMessage());
      return "redirect:/items";
    }

  }

  @GetMapping("/items/{id}/delete")
  public String deleteItem(@PathVariable("id") long id, RedirectAttributes redirectAttributes) {
    try {
      itemRepository.deleteById(id);

      redirectAttributes.addFlashAttribute("message", "The Item with id=" + id + " has been deleted successfully!");
    } catch (Exception e) {
      redirectAttributes.addFlashAttribute("message", e.getMessage());
    }

    return "redirect:/are_doko/home";
  }
}

File: src/main/kotlin/com/kotorisystems/AreDoko/controller/ItemController.kt
package com.kotoriSystems.controller;

import kotlin.collections.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

import com.kotoriSystems.entity.Item;
import com.kotoriSystems.model.SearchForm;
import com.kotoriSystems.repository.ItemRepository;

@Controller
class ItemController(private var itemRepository : ItemRepository) {

  @GetMapping("/items/new")
  fun addItem(@ModelAttribute item : Item) : String {
    return "home_form"
  }

  @PostMapping("/items/save")
  fun saveItem(@ModelAttribute item : Item, redirectAttributes : RedirectAttributes) : String {
    try {
      println("testtest saveItem name" + item.name + "," + item.location?.name + "," + item.
      locationDetail + "," + item.genre?.id + "," + item.genre?.name)
      val existingItem = itemRepository.findById(item.id)
          .orElseThrow { IllegalArgumentException("Invalid item Id:" + item.id) }
      existingItem.name = item.name
      itemRepository.save(existingItem)
      redirectAttributes.addFlashAttribute("message", "The Item has been saved successfully!")
    } catch (e : Exception) {
      println("testtest exception:" + e.message)
      redirectAttributes.addAttribute("message", e.message)
    }

    return "redirect:items"
  }

  @GetMapping("/items")
  fun listItem(@ModelAttribute form : SearchForm) : String {
    val items = itemRepository.findAll()
    for (item in items) {
      println("testtest item:" + item.id + "," + item.name + "," + item.location?.name + "," + item.locationDetail + "," + item.genre?.name)
    }
    form.items = items
    return "home"
  }

  @GetMapping("/items/{id}")
  fun updateItem(@PathVariable("id") id : Long, model : Model, redirectAttributes : RedirectAttributes) : String {
    try {
      // idに対応するItemを取得
      val existingItem = itemRepository.findById(id)
          .orElseThrow { IllegalArgumentException("Invalid item Id:" + id) }
      println("testtest existingItem:" + existingItem.id + "," + existingItem.name + "," + existingItem.location?.name + ","
      + existingItem.locationDetail + "," + existingItem.genre?.name)

      model.addAttribute("item", existingItem)
      model.addAttribute("pageTitle", "Edit Item (ID: " + id + ")")

      return "home_form"
    } catch (e : Exception) {
      redirectAttributes.addFlashAttribute("message", e.message)
      return "redirect:/items"
    }

  }

  @GetMapping("/items/{id}/delete")
  fun deleteItem(@PathVariable("id") id : Long, redirectAttributes : RedirectAttributes) : String {
    try {
      itemRepository.deleteById(id)
      redirectAttributes.addFlashAttribute("message", "The Item with id=" + id + " has been deleted successfully!")
    } catch (e : Exception) {
      redirectAttributes.addFlashAttribute("message", e.message)
    }

    return "redirect:/items"
  }
}
